[contain-intrinsic-size] auto-009.html is failing

cathiechen opened the pull request:
#### d8e31024efa9bde474278105ad85152131220b40
<pre>
[contain-intrinsic-size] auto-009.html is failing
<a href="https://bugs.webkit.org/show_bug.cgi?id=249919">https://bugs.webkit.org/show_bug.cgi?id=249919</a>

Reviewed by Oriol Brufau.

According to [1], &quot;last remembered size&quot; is tracked for the two axises independently, and can be invoked independently.
This patch supports &quot;last remembered size&quot; in a single direction. It replaces the m_lastRememberedSize of
m_lastRememberedLogicalWidth and m_lastRememberedLogicalHeight in ElementRareData. The value -1 means there is
no last remembered width/height.

[1] <a href="https://github.com/w3c/csswg-drafts/issues/7529">https://github.com/w3c/csswg-drafts/issues/7529</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-009-expected.txt:
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::removeChild):
* Source/WebCore/dom/Document.cpp:
(WebCore::CallbackForContainIntrinsicSize):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::lastRememberedLogicalWidth const):
(WebCore::Element::lastRememberedLogicalHeight const):
(WebCore::Element::setLastRememberedLogicalWidth):
(WebCore::Element::clearLastRememberedLogicalWidth):
(WebCore::Element::setLastRememberedLogicalHeight):
(WebCore::Element::clearLastRememberedLogicalHeight):
(WebCore::Element::lastRememberedSize const): Deleted.
(WebCore::Element::setLastRememberedSize): Deleted.
(WebCore::Element::clearLastRememberedSize): Deleted.
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.cpp:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::LayoutUnitMarkableTraits::isEmptyValue):
(WebCore::LayoutUnitMarkableTraits::emptyValue):
(WebCore::ElementRareData::lastRememberedLogicalWidth const):
(WebCore::ElementRareData::lastRememberedLogicalHeight const):
(WebCore::ElementRareData::setLastRememberedLogicalWidth):
(WebCore::ElementRareData::setLastRememberedLogicalHeight):
(WebCore::ElementRareData::clearLastRememberedLogicalWidth):
(WebCore::ElementRareData::clearLastRememberedLogicalHeight):
(WebCore::ElementRareData::useTypes const):
(WebCore::ElementRareData::lastRememberedSize const): Deleted.
(WebCore::ElementRareData::setLastRememberedSize): Deleted.
(WebCore::ElementRareData::clearLastRememberedSize): Deleted.
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::explicitIntrinsicInnerWidth const):
(WebCore::RenderBox::explicitIntrinsicInnerHeight const):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::containIntrinsicLogicalWidthType const):
(WebCore::RenderStyle::containIntrinsicLogicalHeightType const):
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateElementRenderer):

Canonical link: <a href="https://commits.webkit.org/260939@main">https://commits.webkit.org/260939@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c2495ccf6c889450cc84a524c3c2e7d178fe5b3c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109998 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19098 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42680 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1441 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119044 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113951 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20564 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10291 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102256 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115744 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15307 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98523 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43527 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97269 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30180 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/85351 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11801 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31520 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12423 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8477 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17791 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51131 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7583 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14218 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Loirooriol commented on Source/WebCore/dom/ElementRareData.h:
```diff
@@ -199,7 +204,9 @@ class ElementRareData : public NodeRareData {
-    RefPtr<ResizeObserverSize> m_lastRememberedSize;
+
+    LayoutUnit m_lastRememberedLogicalWidth { -1 };
+    LayoutUnit m_lastRememberedLogicalHeight { -1 };
```
I can't find any match for `LayoutUnit m_.*-1`, but there are some `optional<LayoutUnit> m_` (and most seem non-negative). So it may be better to continue the pattern? Though I guess it would use more memory, so not sure.

cathiechen replied:
Yes, that's true. Maybe `std::optional<LayoutUnit>` is better, to keep code consistent.

Loirooriol commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -403,7 +403,10 @@ static void CallbackForContainIntrinsicSize(const Vector<Ref<ResizeObserverEntry
+            if (box->style().containIntrinsicWidthType() == ContainIntrinsicSizeType::AutoAndLength)
+                target->setLastRememberedLogicalWidth(LayoutUnit(contentBoxSize->inlineSize()));
+            if (box->style().containIntrinsicHeightType() == ContainIntrinsicSizeType::AutoAndLength)
+                target->setLastRememberedLogicalHeight(LayoutUnit(contentBoxSize->blockSize()));
```
You need to check the writing mode, `containIntrinsicWidthType` is physical and `setLastRememberedLogicalWidth` is logical.

cathiechen replied:
Done!

Loirooriol commented on Source/WebCore/dom/ElementRareData.cpp:
```diff
@@ -40,7 +40,8 @@ struct SameSizeAsElementRareData : NodeRareData {
     void* resizeObserverData;
-    void* resizeObserverSize;
+    LayoutUnit lastRemembedLogicalWidth;
+    LayoutUnit lastRemembedLogicalHeight;
```
I don't know if it's better to be explicit like this, or be more generic like `void* pointers[16];` and just use `LayoutUnit lastRemembedSize[2];` or even `LayoutUnit layoutUnits[2];`. Leaving this for other reviewers.

cathiechen replied:
Hmm, yes, `LayoutUnit lastRemembedSize[2]` seems better if we use `LayoutUnit`.

Loirooriol commented on Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
```diff
@@ -348,8 +348,12 @@ void RenderTreeUpdater::updateElementRenderer(Element& element, const Style::Ele
+        if (elementUpdateStyle.containIntrinsicWidthType() != ContainIntrinsicSizeType::AutoAndLength)
+            elementUpdateStyle.isHorizontalWritingMode() ? element.setLastRememberedLogicalWidth(LayoutUnit(-1)) : element.setLastRememberedLogicalHeight(LayoutUnit(-1));
+        if (elementUpdateStyle.containIntrinsicHeightType() != ContainIntrinsicSizeType::AutoAndLength)
+            elementUpdateStyle.isHorizontalWritingMode() ? element.setLastRememberedLogicalHeight(LayoutUnit(-1)) : element.setLastRememberedLogicalWidth(LayoutUnit(-1));
```
Setting `-1` will force the creation of the rare data.

cathiechen replied:
Done

cathiechen reviewed (commented):
Thanks @Loirooriol !

Loirooriol commented on Source/WebCore/dom/Element.h:
```diff
@@ -700,9 +699,12 @@ class Element : public ContainerNode {
+    void setLastRememberedLogicalWidth(std::optional<LayoutUnit>);
+    void setLastRememberedLogicalHeight(std::optional<LayoutUnit>);
+    void clearLastRememberedLogicalWidth();
+    void clearLastRememberedLogicalHeight();
```
I think its a bit strange to expose both `set*(std::optional)` and `clear*()`. I would either have

```cpp
void ElementRareData::setLastRememberedLogicalWidth(std::optional<LayoutUnit> width)
{
    m_lastRememberedLogicalWidth = width;
}

void Element::setLastRememberedLogicalWidth(LayoutUnit width)
{
    ensureElementRareData().setLastRememberedLogicalWidth(width);
}

void Element::clearLastRememberedLogicalWidth()
{
    if (auto* data = rareData())
        data->setLastRememberedLogicalWidth(std::nullopt);
}
```

or

```cpp
void ElementRareData::setLastRememberedLogicalWidth(LayoutUnit width)
{
    m_lastRememberedLogicalWidth = width;
}

void ElementRareData::clearLastRememberedLogicalWidth()
{
    m_lastRememberedLogicalWidth.reset();
}

void Element::setLastRememberedLogicalWidth(LayoutUnit width)
{
    ensureElementRareData().setLastRememberedLogicalWidth(width);
}

void Element::clearLastRememberedLogicalWidth()
{
    if (auto* data = rareData())
        data->clearLastRememberedLogicalWidth();
}
```

or

```cpp
void ElementRareData::setLastRememberedLogicalWidth(std::optional<LayoutUnit> width)
{
    m_lastRememberedLogicalWidth = width;
}

void Element::setLastRememberedLogicalWidth(std::optional<LayoutUnit> width)
{
    if (width)
        ensureElementRareData().setLastRememberedLogicalWidth(width);
    else if (auto* data = rareData())
        data->setLastRememberedLogicalWidth(std::nullopt);
}
```

cathiechen replied:
Done, I prefer to the last one:)

cathiechen replied:
Now that we use Markable<LayoutUnit>, I think we could reuse `clear*`, so that the outside don't have to use the magic number(-1).

Loirooriol commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5624,8 +5624,8 @@ std::optional<LayoutUnit> RenderBox::explicitIntrinsicInnerWidth() const
     if (element() && style().containIntrinsicWidthType() == ContainIntrinsicSizeType::AutoAndLength && shouldSkipContent()) {
-        if (auto lastRememberedSize = element()->lastRememberedSize())
-            return isHorizontalWritingMode() ? LayoutUnit(lastRememberedSize->inlineSize()) : LayoutUnit(lastRememberedSize->blockSize());
+        if (auto width = isHorizontalWritingMode() ? element()->lastRememberedLogicalWidth() : element()->lastRememberedLogicalHeight(); width.has_value())
```
I don't think you need `width.has_value()` since `std::optional` has `operator bool()`.

cathiechen replied:
Done

Loirooriol reviewed (approved):
In general it looks good, but I'm not sure what's the policy on enlarging SameSizeAsElementRareData, so better wait for another reviewer to confirm.

smfr commented on Source/WebCore/dom/Element.h:
```diff
@@ -700,9 +699,10 @@ class Element : public ContainerNode {
-    ResizeObserverSize* lastRememberedSize() const;
-    void setLastRememberedSize(Ref<ResizeObserverSize>&&);
-    void clearLastRememberedSize();
+    std::optional<LayoutUnit> lastRememberedLogicalWidth() const;
```
Please use Markable<> for space efficiency.

cathiechen replied:
Done!

cathiechen replied:
Ah, sorry, I replied the wrong line...
Should we keep `std::optional<LayoutUnit>` here and use `Markable<LayoutUnit>` in `ElementRareData`?
That we can keep -1 in `ElementRareData`, and code outside `ElementRareData` doesn't have to know -1 is the empty value.
Not sure if I explain myself clear, please take a look at the new patch, thanks:)

smfr replied:
You should keep the `std::optional<LayoutUnit>` return value. There's automatic conversion from `Markable<>`.

cathiechen replied:
Done, then this is resolved now:)

smfr commented on Source/WebCore/dom/Element.h:
```diff
@@ -700,9 +699,10 @@ class Element : public ContainerNode {
-    void setLastRememberedSize(Ref<ResizeObserverSize>&&);
-    void clearLastRememberedSize();
+    std::optional<LayoutUnit> lastRememberedLogicalWidth() const;
+    std::optional<LayoutUnit> lastRememberedLogicalHeight() const;
```
Ditto.

smfr commented on Source/WebCore/dom/ElementRareData.cpp:
```diff
@@ -40,7 +40,7 @@ struct SameSizeAsElementRareData : NodeRareData {
     void* typedOMData[2];
     void* resizeObserverData;
-    void* resizeObserverSize;
+    std::optional<LayoutUnit> lastRemembedSize[2];
```
Use Markable<> here too.

cathiechen replied:
Done!

cathiechen reviewed (commented):
@smfr Thanks for the review!

smfr commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -403,7 +404,10 @@ static void CallbackForContainIntrinsicSize(const Vector<Ref<ResizeObserverEntry
             auto contentBoxSize = entry->contentBoxSize().at(0);
-            target->setLastRememberedSize(ResizeObserverSize::create(contentBoxSize->inlineSize(), contentBoxSize->blockSize()));
+            if (box->style().containIntrinsicWidthType() == ContainIntrinsicSizeType::AutoAndLength)
+                box->isHorizontalWritingMode() ? target->setLastRememberedLogicalWidth(LayoutUnit(contentBoxSize->inlineSize())) : target->setLastRememberedLogicalHeight(LayoutUnit(contentBoxSize->blockSize()));
```
It's better to use `if/else` than the trigraph operator for code like this.

cathiechen replied:
Done

Loirooriol commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -403,7 +404,18 @@ static void CallbackForContainIntrinsicSize(const Vector<Ref<ResizeObserverEntry
+                    target->setLastRememberedLogicalHeight(LayoutUnit(contentBoxSize->blockSize()));
+                else
+                    target->setLastRememberedLogicalWidth(LayoutUnit(contentBoxSize->inlineSize()));
+            }
```
Might be more readable if you add

```cpp
ContainIntrinsicSizeType containIntrinsicLogicalWidthType() const { return isHorizontalWritingMode() ? containIntrinsicWidthType() : containIntrinsicHeightType(); }
ContainIntrinsicSizeType containIntrinsicLogicalHeightType() const { return isHorizontalWritingMode() ? containIntrinsicHeightType() : containIntrinsicWidthType(); }
```

and then just

```cpp
if (box->style().containIntrinsicLogicalWidthType() == ContainIntrinsicSizeType::AutoAndLength)
    target->setLastRememberedLogicalWidth(LayoutUnit(contentBoxSize->inlineSize()));
if (box->style().containIntrinsicLogicalHeightType() == ContainIntrinsicSizeType::AutoAndLength) {
    target->setLastRememberedLogicalHeight(LayoutUnit(contentBoxSize->blockSize()));
```

cathiechen replied:
Done, and yes, it looks much better:)

Loirooriol commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -607,6 +607,8 @@ class RenderStyle {
     std::optional<Length> containIntrinsicWidth() const { return m_nonInheritedData->rareData->containIntrinsicWidth; }
+    ContainIntrinsicSizeType containIntrinsicLogicalWidthType() const { return isHorizontalWritingMode() ? containIntrinsicWidthType() : containIntrinsicHeightType(); }
+    ContainIntrinsicSizeType containIntrinsicLogicalHeightType() const { return isHorizontalWritingMode() ? containIntrinsicHeightType() : containIntrinsicWidthType(); }
     std::optional<Length> containIntrinsicHeight() const { return m_nonInheritedData->rareData->containIntrinsicHeight; }
```
Adding this between `containIntrinsicWidth` and `containIntrinsicHeight` seems strange. Better

```cpp
    ContainIntrinsicSizeType containIntrinsicWidthType() const { return static_cast<ContainIntrinsicSizeType>(m_nonInheritedData->rareData->containIntrinsicWidthType); }
    ContainIntrinsicSizeType containIntrinsicHeightType() const { return static_cast<ContainIntrinsicSizeType>(m_nonInheritedData->rareData->containIntrinsicHeightType); }
    ContainIntrinsicSizeType containIntrinsicLogicalWidthType() const { return isHorizontalWritingMode() ? containIntrinsicWidthType() : containIntrinsicHeightType(); }
    ContainIntrinsicSizeType containIntrinsicLogicalHeightType() const { return isHorizontalWritingMode() ? containIntrinsicHeightType() : containIntrinsicWidthType(); }
    std::optional<Length> containIntrinsicWidth() const { return m_nonInheritedData->rareData->containIntrinsicWidth; }
    std::optional<Length> containIntrinsicHeight() const { return m_nonInheritedData->rareData->containIntrinsicHeight; }
```

cathiechen replied:
Ah, yes, that's true...

cathiechen committed:
[contain-intrinsic-size] auto-009.html is failing
https://bugs.webkit.org/show_bug.cgi?id=249919

Reviewed by Oriol Brufau.

According to [1], "last remembered size" is tracked for the two axises independently, and can be invoked independently.
This patch supports "last remembered size" in a single direction. It replaces the m_lastRememberedSize of
m_lastRememberedLogicalWidth and m_lastRememberedLogicalHeight in ElementRareData. The value -1 means there is
no last remembered width/height.

[1] https://github.com/w3c/csswg-drafts/issues/7529

* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-009-expected.txt:
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::removeChild):
* Source/WebCore/dom/Document.cpp:
(WebCore::CallbackForContainIntrinsicSize):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::lastRememberedLogicalWidth const):
(WebCore::Element::lastRememberedLogicalHeight const):
(WebCore::Element::setLastRememberedLogicalWidth):
(WebCore::Element::clearLastRememberedLogicalWidth):
(WebCore::Element::setLastRememberedLogicalHeight):
(WebCore::Element::clearLastRememberedLogicalHeight):
(WebCore::Element::lastRememberedSize const): Deleted.
(WebCore::Element::setLastRememberedSize): Deleted.
(WebCore::Element::clearLastRememberedSize): Deleted.
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.cpp:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::LayoutUnitMarkableTraits::isEmptyValue):
(WebCore::LayoutUnitMarkableTraits::emptyValue):
(WebCore::ElementRareData::lastRememberedLogicalWidth const):
(WebCore::ElementRareData::lastRememberedLogicalHeight const):
(WebCore::ElementRareData::setLastRememberedLogicalWidth):
(WebCore::ElementRareData::setLastRememberedLogicalHeight):
(WebCore::ElementRareData::clearLastRememberedLogicalWidth):
(WebCore::ElementRareData::clearLastRememberedLogicalHeight):
(WebCore::ElementRareData::useTypes const):
(WebCore::ElementRareData::lastRememberedSize const): Deleted.
(WebCore::ElementRareData::setLastRememberedSize): Deleted.
(WebCore::ElementRareData::clearLastRememberedSize): Deleted.
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::explicitIntrinsicInnerWidth const):
(WebCore::RenderBox::explicitIntrinsicInnerHeight const):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::containIntrinsicLogicalWidthType const):
(WebCore::RenderStyle::containIntrinsicLogicalHeightType const):
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateElementRenderer):

Canonical link: https://commits.webkit.org/260939@main
